### PR TITLE
issue #3495 - more selectively scope searches to patient compartment

### PR DESCRIPTION
**Pull request by lmsurpre**
Specifically, avoid adding the implicit compartment parameter(s) when
the interaction is allowed by one or more scopes with the 'user/' or
'system/' prefix ("context type")

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 1]
```diff
@@ -695,51 +713,50 @@ private void enforce(String resourceType, String resourceId, Resource resource,
...
-            return true;
-        }
 
-        if (approvedScopeMap.containsKey(ContextType.USER)) {
-            // For `user` scopes, we grant access to all resources of the requested type.
-            // Implementers that use these scopes are encouraged to layer on their own permissions model beyond this.
-            if (log.isLoggable(Level.FINE)) {
-                log.fine(requiredPermission.value() + " permission for '" + resourceType + "/" + resourceId +
-                    "' is granted via scope " + approvedScopeMap.get(ContextType.USER));
+            Optional<Scope> approvingScope = grantedScopes.get(context).stream()
```
Comment (punktilious) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:732]: comment here would be useful on what this section is intending to achieve

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 2]
```diff
@@ -695,51 +713,50 @@ private void enforce(String resourceType, String resourceId, Resource resource,
...
+                    .filter(s -> s.getResourceType() == ResourceType.Value.RESOURCE || s.getResourceType().value().equals(resourceType))
+                    .filter(s -> hasPermission(s.getPermission(), requiredPermission))
+                    .findAny();
+            if (approvingScope.isPresent()) {
+                if (log.isLoggable(Level.FINE)) {
```
Comment (punktilious) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:737]: I noticed that in other places we're not logging which scope resulted in the request being approved. Perhaps approvals should be logged more broadly with FINER?

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 3]
```diff
@@ -695,51 +713,50 @@ private void enforce(String resourceType, String resourceId, Resource resource,
...
 
-        if (approvedScopeMap.containsKey(ContextType.PATIENT)) {
+        if (grantedScopes.containsKey(ContextType.PATIENT)) {
+            Optional<Scope> approvingScope = grantedScopes.get(ContextType.PATIENT).stream()
+                    .filter(s -> s.getResourceType() == ResourceType.Value.RESOURCE || s.getResourceType().value().equals(resourceType))
+                    .filter(s -> hasPermission(s.getPermission(), requiredPermission))
+                    .findAny();
+
             if (resource == null) {
                 // For `patient` scopes, if no resource was retrieved, which happens with history/search
```
Comment (punktilious) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:752]: what about deleted resources? The resource value will be null for those too

**Review by punktilious — COMMENTED** [review]
Also in the file, `therefor` should be `therefore`

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 1 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:732]: will add a comment like this:
> // look for any scope that approves the current interaction

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:752]: we don't call this from `afterDelete`, but we do call it from `afterHistory` and so I agree this is worth testing...I'll give it a try and maybe add a test

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:737]: We don't have a good "single spot" to add this logging, but I could add it to the "isApprovedByScopes" method which is used from a number of thee `beforeX` methods.  I think log level FINE is ok for these and would be consistent with the other spots.

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:752]: this wasn't really from this PR, but instead from https://github.com/IBM/FHIR/issues/3423
however, i confirmed that, with those changes, we now reject a history interaction if any one of the resource versions in the history was a delete...i'll see if i can address that in this PR.  if not, i'll open a separate issue for it

**Review by tbieste — APPROVED** [review]

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 1 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:834]: added

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:839]: added

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java]: I made some changes to try addressing this one.  please have another look

**Review by punktilious — APPROVED** [review]
LGTM

**Pull request merged**